### PR TITLE
Add Statisfiability Problem

### PR DESCRIPTION
**Pull request by SciCodePhy**
The Satisfiability model is mostly based on the https://github.com/QuEraComputing/GenericTensorNetworks.jl/blob/master/src/networks/Satisfiability.jl. The following functions are added:
* is_kSAT function for CNF and Satisfiability class
* length for CNF clause
* variables function
* terms function

### CODE REVIEW

**Comment by SciCodePhy** [comment]
Fix test/Satisfiability file

File: src/models/Satisfiability.jl [hunk 1]
```diff
@@ -167,12 +159,29 @@ function variables(c::Satisfiability)
...
     end
-    return collect(var_names)  # Convert the set back to a vector if needed
+    return collect(var_names) 
 end
 flavors(::Type{<:Satisfiability}) = [0, 1]  # false, true
 terms(c::Satisfiability) = collect(c.cnf.clauses)
 parameters(sat::Satisfiability) = Int[]
 
+function is_kSAT(x)
+    if isa(x, Satisfiability)
```
Comment (hmyuuu) [inline src/models/Satisfiability.jl:169]: How about we dispatch, e. g.
```julia
is_kSAT(s::Satisfiability) = is_kSAT(s.cnf)
```

Comment (GiggleLiu) [reply src/models/Satisfiability.jl:169]: I am not sure we need this function.

File: src/models/Satisfiability.jl [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    var_names = Set{}()
+    for clause in c.cnf.clauses
+        for var in clause.vars
+            push!(var_names, var.name)
+        end
+    end
+    return collect(var_names) 
+end
+flavors(::Type{<:Satisfiability}) = [0, 1]  # false, true
+terms(c::Satisfiability) = collect(c.cnf.clauses)
```
Comment (GiggleLiu) [inline src/models/Satisfiability.jl:165]: NOTE: we may consider removing the `terms` interface. This is only required for energy based models.

Comment (GiggleLiu) [reply src/models/Satisfiability.jl:165]: Please check #38

File: src/models/Satisfiability.jl [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+"""
+function satisfiable(cnf::CNF{T}, config::AbstractDict{T}) where T
+    all(x->satisfiable(x, config), cnf.clauses)
+end
+function satisfiable(c::CNFClause{T}, config::AbstractDict{T}) where T
+    any(x->satisfiable(x, config), c.vars)
+end
+function satisfiable(v::BoolVar{T}, config::AbstractDict{T}) where T
+    config[v.name] == ~v.neg
+end
```
Comment (GiggleLiu) [inline src/models/Satisfiability.jl:198]: Please add an `evaluate` function.

File: src/models/Satisfiability.jl (hunk 1 above)
Comment (SciCodePhy) [reply src/models/Satisfiability.jl]: This function might be needed for the reduction from k-SAT to 3-SAT.

File: src/models/Satisfiability.jl [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    print(io, join(["($k)" for k in c.clauses], " ∧ "))
+end
+Base.:(==)(x::CNF, y::CNF) = x.clauses == y.clauses
+Base.length(x::CNF) = length(x.clauses)
+function is_kSAT(cnf::CNF)
+    len_lst = [length(clause.vars) for clause in cnf.clauses]
+    if length(unique(len_lst)) == 1
+        return len_lst[1]
+    else
+        return false
```
Comment (GiggleLiu) [inline src/models/Satisfiability.jl:60]: Return value not type stable: https://docs.julialang.org/en/v1/manual/performance-tips/

**Comment by GiggleLiu** [comment]
```julia
struct KSatisfiability{K, T} <:AbstractProblem
    cnf::CNF{T}
    function KSatisfiability(cnf::CNF{T}) where {T}
        @assert is_kSAT(cnf) == K "The CNF is not a $K-SAT problem"
        new{K, T}(cnf)
    end
end
```

**Review by GiggleLiu — APPROVED** [review]
Please add more tests to improve the test coverage.

**Comment by SciCodePhy** [comment]
[WIP] More tests will be added to the test file to increase the coverage.

**Pull request merged**